platform/ext/target/nordic_nrf/common/native_drivers/spu.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "spu.h"
+#include <hal/nrf_spu.h>
+#include "region_defs.h"
+
+/* Platform-specific configuration */
+#define FLASH_SECURE_ATTRIBUTION_REGION_SIZE SPU_FLASH_REGION_SIZE
+#define SRAM_SECURE_ATTRIBUTION_REGION_SIZE  SPU_SRAM_REGION_SIZE
+
+#define NUM_FLASH_SECURE_ATTRIBUTION_REGIONS \
+	(FLASH_TOTAL_SIZE / FLASH_SECURE_ATTRIBUTION_REGION_SIZE)
+#define NUM_SRAM_SECURE_ATTRIBUTION_REGIONS \
+	(TOTAL_RAM_SIZE / SRAM_SECURE_ATTRIBUTION_REGION_SIZE)
+
+
+/* Convenience macros for SPU Non-Secure Callable (NCS) attribution */
+
+/*
+ * Determine the SPU Region number the given address belongs to.
+ *
+ * addr shall be a valid flash memory address
+ */
+#define FLASH_NSC_REGION_FROM_ADDR(addr) \
+	((u32_t)addr / FLASH_SECURE_ATTRIBUTION_REGION_SIZE)
+
+/*
+ * Determine the NSC region size based on a given NCS region base address.
+ */
+#define FLASH_NSC_SIZE_FROM_ADDR(addr) (FLASH_SECURE_ATTRIBUTION_REGION_SIZE \
+	- (((u32_t)(addr)) % FLASH_SECURE_ATTRIBUTION_REGION_SIZE))
+
+/*
+ * Determine the encoded the SPU NCS Region Size value,
+ * based on the absolute NCS region size in bytes.
+ */
+#define FLASH_NSC_SIZE_REG(size) ((31 - __builtin_clz(size)) - 4)
```

Comment:
I think this gives the wrong value if size is a power of 2 and if it is less than 16.
We want
`FLASH_NSC_SIZE_REG(0) == FLASH_NSC_SIZE_REG(15) == FLASH_NSC_SIZE_REG(31) == FLASH_NSC_SIZE_REG(32) == 1`
and
`FLASH_NSC_SIZE_REG(33) == FLASH_NSC_SIZE_REG(64) == 2` etc.
https://infocenter.nordicsemi.com/topic/ps_nrf9160/spu.html?cp=2_0_0_5_14_8_16#register.FLASHNSC-0-1.SIZE
```suggestion
#define FLASH_NSC_SIZE_REG(size) (33 - __builtin_clz((size - 1)>>5))
```
`size - 1` fixes for powers of 2, while the other changes fix for less than 16.
We should probably refer to the PS in the comment otherwise this macro will be confusing for readers :P.

Reply:
@oyvindronningstad @anangl may we, instead, agree that this is the responsibility of the linker? This is how it is done in SPM as well. Also, this is how the function is documented:

```
 If the start address of the NSC area is hard-coded, it must follow
 * the HW restrictions: The size must be a power of 2 between 32 and
 * 4096, and the end address must fall on a SPU region boundary.
```
I think all the above  should be checked with proper ASSERTS. I added an ASSERT for the range checks.

I also enhanced the description of the macro
```
/*
 * Determine the encoded the SPU NCS Region Size value,
 * based on the absolute NCS region size in bytes.
 *
 * size shall be a valid SPU NCS Region size value
 */

so it now warns that we need to pass a valid size.

I don't want this macro to do truncation left or right, I want this macro to only do th encoding.
#define FLASH_NSC_SIZE_REG(size) ((31 - __builtin_clz(size)) - 4)
```

Reply:
Yep, sorry, I forgot that the parameter has been checked before calling this macro.

I guess we should provide the linker changes to tfm_common_s.ld.template, will you do that in this PR?